### DISCUSSION TITLE
How do you read a file line by line?

### DISCUSSION THREAD

**Discussion by MrMidnightOfficial** [vlang/v#14840] [General]
How do you read a file line by line?

**Comment by JalonSolov** [comment]
You don't.  Not as a simple library call, at least.

You can read all the lines from a file into an array of strings, one array entry per line of the file.

You can read all the bytes of a file into a byte array, and step through it yourself to find EOL characters.

You can use a buffered reader to read bytes from a file until you hit an EOL, then save that as a string.

Etc.

**Comment by medvednikov** [comment]
File streams are planned. They are necessary for very large files.

**Reply by JalonSolov** [reply]
Yep.  Just not there right now.

**Comment by Wajinn** [comment]
Possibly the OP is referring to `os.read_lines` (https://modules.vlang.io/os.html#read_lines) and `os.read_file` (https://modules.vlang.io/os.html#read_file), using `import os` .  

`os.read_lines` of course works.  But, if the file is extremely large, have encountered some issues, though you could of course use `os.read_file in many such cases.

For the sake of the OP...  A simple example of possible usage would kind of be like below.

```
mut view_content := []string{}

if os.is_file(path) == true // check if file exists
{
	view_content << os.read_lines(path) or {panic('not read: $path')}
}
```

**Reply by JalonSolov** [reply]
Reading all the lines in a file at once into an array is different from reading a file line by line.

**Reply by Wajinn** [reply]
Yes, you are correct.  Just taking a guess at what the OP might have meant, was possibly playing with, and throwing some stuff on the wall.

**Comment by hholst80** [comment]
Just in case anyone finds this, this has been implemented.

```v

import os

fn main() {
    file_path := 'path/to/your/file.txt'
    mut file := os.open(file_path) or {
        eprintln('Failed to open the file: $err')
        return
    }
    defer { file.close() }  // Automatically close the file when the function returns

    // Read lines from the file
    for {
        line := file.read_line() or { break }
        println(line)
    }
}


```

**Reply by LukyGuyLucky** [reply]
May I ask is this already workable?I tried yesterday after a fresh update of V but failed to compile.

**Reply by hholst80** [reply]
Sorry! I did not test the code. It turns out that the API I thought I saw was not on File but on BufferedReader.  Which makes perfect sense, and the way you would implement a line parser anyway.

https://github.com/vlang/v/blob/master/examples/buf_reader.v

```v
import time
import io
import os

fn main() {
        file := os.open("/etc/fstab")!
	mut r := io.new_buffered_reader(reader: file)
	for {
		line := r.read_line() or { break }
		println(line)
		// Make it nice and obvious that we are doing this line by line
		time.sleep(100 * time.millisecond)
	}
}
```

**Reply by Jengro777** [reply]
If I want to write the content to a file line by line, how can I automatically create the file if the file does not exist?

**Comment by LukyGuyLucky** [comment]
Well noted.Thanks for the reply!
________________________________
发件人: Henrik Holst ***@***.***>
发送时间: 2023年12月28日 8:42
收件人: vlang/v ***@***.***>
抄送: LukyGuyLucky ***@***.***>; Comment ***@***.***>
主题: Re: [vlang/v] How do you read a file line by line? (Discussion #14840)


Sorry! I did not test the code. It turns out that the API I thought I saw was not on File but on BufferedReader:

https://github.com/vlang/v/blob/master/examples/buf_reader.v

―

**Comment by myastrotong** [comment]
vlang can directly call C code, so you can use C fgets() function in v:
vlang is fundamentally a better C, so it is not a shame to directlly use C in v! 
below is a demo:
```
#include <stdio.h>

fn main(){
	file_name :="readfile_linebyline.v"	
	fp:=C.fopen(&char(file_name.str),c'rb')
	if 0==fp {
		eprintln('open file err!' )  
	}
	defer{
		C.fclose(fp)
	}
	
	mut line :=''.repeat(1024)	
	for{
		ret :=C.fgets(line.str,1024,fp)
		
		unsafe{
			print(line.str.vstring())
		}
		
		if 0==ret {
			break
		}
	}
}
```
